src/palimpzest/utils/progress.py

```diff
@@ -47,179 +40,230 @@ def get_memory_usage() -> float:
+        pass
+
+    @abstractmethod
+    def incr(self, op_id: str, num_outputs: int = 1, display_text: str | None = None, **kwargs):
+        """
+        Advance the progress bar for the given operator by one. Modify the downstream operators'
+        progress bar `total` to reflect the number of outputs produced by this operator.
+        """
+        pass
+
+    @abstractmethod
+    def finish(self):
+        """Clean up and finalize progress tracking"""
+        pass
+
+    def update_stats(self, op_id: str, **kwargs):
+        """Update progress statistics"""
+        for key, value in kwargs.items():
+            if hasattr(self.op_id_to_stats[op_id], key):
+                if key != "total_cost":
+                    setattr(self.op_id_to_stats[op_id], key, value)
+                else:
+                    self.op_id_to_stats[op_id].total_cost += value
+        self.op_id_to_stats[op_id].memory_usage_mb = get_memory_usage()
+
+
+class MockProgressManager(ProgressManager):
+    """Mock progress manager for testing purposes"""
+
+    def __init__(self, plan: PhysicalPlan | SentinelPlan, num_samples: int | None = None):
+        pass
+
+    def add_task(self, op_id: str, op_str: str, total: int):
+        pass
+
+    def start(self):
+        pass
+
+    def incr(self, op_id: str, num_outputs: int = 1, display_text: str | None = None, **kwargs):
+        pass
+
+    def finish(self):
+        pass
+
+
+class PZProgressManager(ProgressManager):
+    """Progress manager for command line interface using rich"""
+    
+    def __init__(self, plan: PhysicalPlan | SentinelPlan, num_samples: int | None = None):
+        super().__init__(plan, num_samples)
+        self.console = Console()
+
+    def add_task(self, op_id: str, op_str: str, total: int):
+        """Add a new task to the progress bar"""
+        task = self.progress.add_task(
+            f"[blue]{op_str}", 
             total=total,
             cost=0.0,
             success=0,
             failed=0,
             memory=0.0,
-            recent=""
+            recent="",
         )
-        
-        # Start progress bar
+
+        # store the mapping of operator ID to task ID
+        self.op_id_to_task[op_id] = task
+
+        # initialize the stats for this operation
+        self.op_id_to_stats[op_id] = ProgressStats(start_time=time.time())
+
+    def start(self):
+        # print a newline before starting to separate from previous output
+        print()
+
+        # set start time
+        self.start_time = time.time()
+
+        # start progress bar
         self.progress.start()
-        
-    def update(self, current: int, sample: str | None = None, **kwargs):
-        self.update_stats(**kwargs)
-        
-        # Update progress bar and recent text in one update
-        if sample:
-            self.stats.recent_text = sample
-            
+
+    def incr(self, op_id: str, num_outputs: int = 1, display_text: str | None = None, **kwargs):
```

Comment:
Can you add a comment about how we expect num_outputs to be 1. If its not 1, we will update the total of downstream operators at the declaration of incr in the base class.